File: c152h.rb
# reddit.com/r/DailyProgrammer
# Solution to Challenge #???: Minimum Spanning Tree

vertices = gets.chomp.to_i
adjacency = Array.new(vertices) { gets.chomp.split(',').map {|n| n.to_i } }.transpose # matrix input one liner

traversed_vertices = [0]
edges = []
alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

# using Prim's algorithm
while traversed_vertices.count < vertices do
  possible_verts = adjacency.map.with_index do |col, y|
    {column: col, index: y } # get indices of columns
  end.select do |col|
    traversed_vertices.include? col[:index] # exclude columns not traversed yet
  end.map do |col|
    col[:column].map.with_index do |edge_weight, x|
      {row: x, col: col[:index], weight: edge_weight} # collate edge data
    end.select do |vert|
      not traversed_vertices.include? vert[:row] # exclude rows already traversed
    end
  end.flatten
  possible_verts.select! {|vert| vert[:weight] >= 0} # get rid of -1s (ie. no edge)
  possible_verts.sort_by! {|vert| vert[:weight]} # sort
  best_match = possible_verts[0] # get shortest
  edges.push best_match
  traversed_vertices.push best_match[:row]
end

weight = edges.map {|edge| edge[:weight]}.inject {|wt, v| wt + v}
edges.map! {|edge| alphabet[edge[:col]] + alphabet[edge[:row]] }
puts weight
puts edges.join ','

File: c152h.clj
; Input via stdin
; Formatted as described in reddit.com/r/dailyprogrammer/comments/20cydp/14042014_challenge_152_hard_minimum_spanning_tree/

(defn with-index [coll]
  (map-indexed vector coll))

(defn get-adjacency []
  (with-index (map with-index
                   (repeatedly (read-string (read-line))
                               #(map read-string (clojure.string/split (read-line) #", *"))))))

(defn vert-from [adj indices bool-func]
  (filter (fn [current-index] (bool-func (some (fn [test-index] (= (first current-index) test-index))
                                               indices)))
          adj))

(defn get-edges [coll]
  (filter (fn [kv] (>= (get kv 1) 0))
          coll))

(defn from-adjacency [adj current]
  (apply concat
         (map (fn [il] (map (fn [vert] (conj vert (first il)))
                            (vert-from (last il) current identity)))
              (vert-from adj current not))))

(defn prim [adj current]
  (if (< (count current) (count adj))
    (let [next-vert (apply min-key
                           (fn [vert] (get vert 1))
                           (get-edges (from-adjacency adj current)))]
      (concat [next-vert]
              (prim adj (conj current (last next-vert)))))
    nil))

(let [tree (prim (get-adjacency) [0])]
  (println (apply + (map (fn [vert] (get vert 1)) tree)))
  (println (let [alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZ"]
             (clojure.string/join ","
                                  (map (fn [vert] (str (get alphabet (first vert))
                                                       (get alphabet (last vert))))
                                       tree)))))
